Votes - model and spec

gingin77 opened the pull request:
This is the beginning of a votes feature that allows authenticated users to vote on both posts and comments

gingin77 committed:
Created vote model and specs, existing tests are all passing

gingin77 committed:
Add more specs for Vote model and more validations.

gingin77 committed:
Indentation adjustments after trying out a linter for Rubocop

gingin77 committed:
Remove spec for votes controller - it is on another branch

gingin77 committed:
Adjust indentations and spacing in Vote model and in spec for vote model

zporter commented on app/models/vote.rb:
```diff
@@ -0,0 +1,17 @@
+class Vote < ApplicationRecord
+  belongs_to :voter,
+             class_name:  :User,
```
I've never seen a symbol passed in before. neat!

zporter commented on app/models/vote.rb:
```diff
@@ -0,0 +1,17 @@
+                                      only_integer:             true,
+                                      greater_than_or_equal_to: -1,
+                                      less_than_or_equal_to:    1
+                                    }
```
not super critical or anything, but both of these validations could've just used the `validates` method like so:

```ruby
class Vote < ApplicationRecord
  validates :user_id,
            uniqueness: { scope: [:voteable_id, :voteable_type] }

  validates :value,
            numericality: {
              only_integer:             true,
              greater_than_or_equal_to: -1,
              less_than_or_equal_to:    1
            }
end
```

Saves some with indentation.

Another thing is that the numericality validation could be replaced with an inclusion. something like:

```ruby
validates :value, inclusion: { in: [1, 0, -1] }
```

zporter commented on config/application.rb:
```diff
@@ -14,6 +14,7 @@ class Application < Rails::Application
     config.generators do |g|
       g.assets false
       g.helper false
+      g.test_framework  :rspec, fixture: false
```
extra space between `test_framework` and `:rspec`

zporter commented on db/migrate/20180216204115_create_votes.rb:
```diff
@@ -0,0 +1,10 @@
+      t.references :user, null: false
+      t.integer :value, null: false
+      t.timestamps
+    end
```
might be good to add a uniqueness constraint here as well.

```ruby
create_index :votes, [:user_id, :voteable_id, :voteable_type], unique: true
```

remove the `index: true` on the `voteable` references if you end up going that route.

zporter commented on spec/models/voter_spec.rb:
```diff
@@ -0,0 +1,22 @@
+  it { is_expected.to have_attribute(:value) }
+
+  it { is_expected.to belong_to(:voteable) }
+  it { is_expected.to belong_to(:voter) }
```
might also be good to tack on the `class_name` and `foreign_key` options on here as well.

```ruby
it { is_expected.to belong_to(:voter).class_name(User).with_foreign_key(:user_id) }
```

zporter commented on spec/models/voter_spec.rb:
```diff
@@ -0,0 +1,22 @@
+  it do
+    is_expected.to validate_uniqueness_of(:user_id)
+      .scoped_to(:voteable_id, :voteable_type)
+  end
```
nitpick: add a space after this block.

gingin77 replied:
Added!

zporter commented on spec/support/factories/votes.rb:
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :vote do
+    value 1
```
could be cool to randomize the values here with `[1, 0, -1].sample` and setup traits for `:upvote`, `:downvote` and `:novote`

gingin77 replied:
I think I get where you're going with this but I don't know exactly what you mean on a detailed level. Since I don't want to use zero in my validations (there's a null constraint on the value attribute at the db table level), I'm only dealing with either 1 or -1. I created traits for up_vote and down_vote, which may be useful for future feature specs.

zporter replied:
FWIW, `0` is not null. I also pulled those 3 numbers because that's what the numerical validation implied before the changes. `[1, -1].sample` would be the correct update now that I see all the values that `value` can be. I just thought it'd be cool to still be able to say `build :vote` if you didn't care what the `value` was. The traits look great 👍

gingin77 replied:
Gotcha - I'm making an update now to say:
`FactoryBot.define do`
  `    factory :vote do`
    `value [1, -1].sample`

zporter reviewed (changes requested):
add the uniqueness constraint on the db, please and thank you.

gingin77 committed:
Remove extra space

gingin77 committed:
Update indents and spacing in Vote model;

Revisit models for Post and Comments to make all 3 models use similar
layout.

gingin77 committed:
Add uniqueness constaint to Vote table

gingin77 committed:
For Vote specs - improve votes factory by adding traits

In vote_spec: add class_name and foreign key reference to belongs_to
test; make formatting changes

zporter commented on app/models/comment.rb:
```diff
@@ -1,8 +1,14 @@
 
-  validates :text, presence: { message: "The comment you submitted was blank" }
+  has_many :comments,
+           as: :commentable
```
what do you think about a `dependent: :destroy` being set on this association?

zporter commented on app/models/post.rb:
```diff
@@ -1,9 +1,21 @@
-  validates :title, :url, presence: true, if: :url_post?
-  validates :title, :url, absence: true, if: :text_post?
+  has_many :comments,
+           as: :commentable
```
what do you think about a `dependent: :destroy` being set on this association?

zporter commented on app/models/post.rb:
```diff
@@ -1,9 +1,21 @@
+           as: :commentable
+
+  has_many :votes,
+           as: :voteable
```
what do you think about a `dependent: :destroy` being set on this association?

zporter commented on app/models/user.rb:
```diff
@@ -6,4 +6,5 @@ class User < ApplicationRecord
 
   has_many :posts
   has_many :comments
+  has_many :votes
```
what do you think about a `dependent: :destroy` being set on all of these associations?

zporter commented on app/models/vote.rb:
```diff
@@ -0,0 +1,16 @@
+  validates :value,
+            presence: true,
+            numericality: true,
+            inclusion: { in: [1, -1] }
```
this is super neat, but I'm curious whether the `numericality: true` is adding any value here on top of the `inclusion` check 🤔

gingin77 replied:
Only in a circular way - a test for numericality was failing when `numericality: true` was absent. When I was first writing the tests, I started with deciding that `value` needed to be a number. I prefer to strictly allow only 1 or negative 1. I'm going to take out the numericality test and this line.

zporter commented on db/migrate/20180216204115_create_votes.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateVotes < ActiveRecord::Migration[5.1]
+  def change
+    create_table :votes do |t|
+      t.references :voteable, polymorphic: true
```
would it be good for this to also be `null: false`?

zporter commented on spec/models/voter_spec.rb:
```diff
@@ -0,0 +1,26 @@
+
+  it { is_expected.to have_attribute(:value) }
+
+  it { is_expected.to belong_to(:voteable) }
```
nitpick: add space after this line since it's switching to a block-level (`it do`) syntax.

zporter commented on spec/support/factories/votes.rb:
```diff
@@ -0,0 +1,18 @@
+    end
+
+    factory :down_vote, traits: [:down_vote]
+    factory :up_vote, traits: [:up_vote]
```
huh, I didn't know this was possible.  Super neat!

gingin77 replied:
Cool - but I didn't invent anything here - just copied what is in the docs :) https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#traits

zporter reviewed (approved):
just a few nitpicks, but nothing to stop from merging this PR. great work 👍

gingin77 committed:
Add dependent destroy to models and specs w/ a belongs-to relationships

gingin77 committed:
Add null false constraint on voteable in db table
